074-k8s-storage/proposal.md

```diff
@@ -0,0 +1,167 @@
+
+# Motivation
+
+As we started thinking about the Kubernetes runtime we realized that we need to think about what our storage solution would be before proceeding with any other part of the implementation. Storage has a huge effect on how Concourse interacts with the runtime (Kubernetes). Storage also had a lot of unknowns, we didn't know what the storage landscape on Kubernetes looked like and what options were available to us. Storage also has a huge impact on the performance of the cluster, in regards to storage and initialization of steps.
+
+## Requirements
+An ideal storage solution can do the following :
+
+- image fetching from the CRI k8s is using
+- transfer **cache objects** between steps (whatever represents a step, most likely a pod)
+- cache for resources and tasks
+- stream **cache objects** across worker runtimes (k8s worker sends artifact to garden worker)
+
+## Criteria
+- security
+- performance, aka initialization time (time spent running workloads on a single k8s worker, as well as across workers)
+- resource usage to run this storage solution
+
+# Proposal
+
+**TL;DR**: We recommend going with the image registry option because it satisfies all the requirements and gives us a bunch of options to improve performance when compared to the blobstore option. It also provides a very flexible solution that works across multiple runtime workers. [See Details](#image-registry-to-store-artifacts)
+
+Furthermore, the CSI is a useful interface for building the storage component against. [See Details](#csi)
+
+# Storage Options considered
+## Baggageclaim Daemonset 
+### Description
+A privileged baggageclaim pod would manage all the **cache object** for step pods. The pod can be provided sufficient privilege to create overlay mounts using `BiDirectional` value for `mountPropagation`. The `volumeMount` object allows specifying a volume `subPath`.
+
+This approach didn't work using GCE PDs or vSphere Volumes ([Issue](https://github.com/kubernetes/kubernetes/issues/95049)). It does work using `hostPath` option, however, that would require a large root volume and wouldn't be able to leverage IaaS based persistent disks. 
+
+The pod would run on all nodes that Concourse would execute steps on.
+ 
+### Pros
++ Leverage baggageclaim
+	+ volume streaming between nodes would work using the current Concourse architecture
+	+ resource and task caches would also work using the current Concourse architecture
+	+ would be able to stream **cache objects** across worker runtimes as it would be mediated via the web
++ Concourse would have complete control over volume lifecycle
++ would have negligible overhead for steps scheduled on the same node as no input/output stream would be required
+
+### Cons
+- Not being able to use IaaS based persisent disks doesn't offer a viable solution. K8s nodes would need to have large root volumes.
```

Comment:
Is this actually an issue? Under `Persistent Volumes` you note as a con:
> IaaS based limits on [volume limits per node](https://kubernetes.io/docs/concepts/storage/storage-limits/#dynamic-volume-limits) prevents this from being a scalable solution
Large root volumes is closer to how Concourse Workers work today (i.e. closer to the implicit foundational design assumptions in the Concourse domain that this RFC attempts to deal with), and if there are scalability issues with Persistent Volumes, then doesn't this represent a simpler way (because it builds on prior `baggageclaim` work) to get started with building the Kubernetes runtime?

Reply:
Hey @ari-becker,
Yes, this is true. It would be useful to get perspectives such as yours for us to understand what constraints or lack thereof folks are dealing with. 

We are assuming the cons;
- resizing of nodes with larger root volumes
- running privileged baggageclaim pods
- being ok with baggagelciam's current lack of security (this can be enhanced)
creates larger challenges for operators.

It sounds like, you'd be okay with this setup ?

Reply:
@xtreme-sameer-vohra currently, we already run Concourse workers on a Kubernetes cluster, one where we've architected our workers such that the workers are restricted to their own underlying nodes (we manually scale both the underlying nodes and the worker StatefulSet in tandem) and the worker pods are privileged. So all of your cons are *already present* in our deployment and therefore implicitly acceptable.

Maybe you need to clarify what the goals are supposed to be of developing a Kubernetes runtime for Concourse? is it to phase out the "worker" in favor of ATC directly scheduling Kubernetes Jobs that represent Concourse jobs or steps? Because then running privileged baggageclaim pods are fine, basically then the worker isn't really phased out, it's just simplified down to only the baggageclaim code, and presumably we'd be able to configure ATC to create Jobs with node restrictions / affinity / taints / tolerations to get the Jobs to run on a dedicated worker instance group. Is it the ability to run on common / shared instances? Because then running privileged baggageclaim pods is probably unacceptable. Should untrusted Concourse workloads run on common / shared instances in the first place, even if they're not allowed to be privileged? Ehhh, debatable, and a lot of security experts will tell you no, particularly if the jobs are not configured with PodSecurityPolicies and the like, which in a CI / development environment rarely makes sense.

The main value that we'd get out of a Kubernetes-native runtime would be the ability to set up [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) so that Kubernetes will add additional nodes (and reduce nodes) as needed according to scheduling workloads, as currently we need to scale workers by hand. Cluster Autoscaler works well with workloads that are scheduled to dedicated instance groups, adding and subtracting nodes from the dedicated instance group that will permit Jobs that are restricted to those instance groups.

Reply:
Ah yes, we are working on the [K8s Runtime RFC](https://github.com/concourse/rfcs/blob/075-k8s-runtime/075-k8s-runtime/proposal.md) that should have preceded this one 😸 
The primary objective being able to leverage the scaling & operational capabilities of K8s by deploying `steps` as K8s pods without changing the pipeline schema. The latter would allow pipelines to be portable across runtimes. Along the same lines we're also assessing;
- the least amount of required permissions on K8s for fulfilling the same requirements as a current worker (container & storage management)
- the # of dependencies (besides postgres) for a Concourse deployment on K8s

As you mentioned, the container management would be delegated to k8s. We are [spiking](https://github.com/concourse/concourse/issues/6133) on using IaaS PVs + propagating mounts using CSI, hoping to ultimately leverage `baggageclaim`. We are early in the exploration, but this path might also provide the ability of using volume location for the purposes of scheduling 💭 

The initial draft of the proposal has a worker being mapped to K8s cluster + namespace. "Instance Groups" seems interesting. Do you see it fitting in naturally with the cluster+namespace mapping or would you propose something different ?

Reply:
@xtreme-sameer-vohra One of the things that worries me about using an external image registry is the prospect of cross-availability-zone (on AWS, $0.01/GB) and/or VPC egress traffic costs (on AWS, much higher). Currently, all of our workers are in the same availability zone (for CI workers, we're less concerned with the prospect of an AZ-wide outage and more concerned with the traffic costs) and so our traffic costs are much lower.

Effectively using a external image registry while keeping costs under control would mean a) deploying a separate image registry per availability zone alongside the workers, where workers are correctly configured to use the image registry that is in their availability zone, b) the image registry itself needs to use either a local disk / IaaS PV (so as not to incur traffic costs when storing layers in s3, for example), so really all that an image registry does is consolidate storage usage in one place. This eventually leads to cost savings compared to using a large local disk / IaaS PV per worker, with the tradeoff of additional operational complexity.

When talking about instance groups, I'm using the [Kops](https://github.com/kubernetes/kops/blob/master/docs/instance_groups.md) nomenclature. An instance group maps to an AWS AutoscalingGroup, where the nodes are of a certain instance type (i.e. CI is generally best-served with burstable nodes, which on AWS are t3 nodes, whereas t3 nodes may not make sense for other workloads). In order to schedule workers on a specific ASG that we set up for them, we use a combination of [node affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity) to direct Kubernetes to schedule the CI workloads (currently, the worker pods) to those nodes, and [taints and tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) to not allow any other workloads to schedule there. This is orthogonal to Kubernetes namespacing, which is a good idea anyway to improve logical isolation between individual CI workloads.